src/identity-providers.js

```diff
@@ -0,0 +1,95 @@
+'use strict'
+const Identity = require('./identity')
+const OrbitDBIdentityProvider = require('./orbit-db-identity-provider')
+const Keystore = require('orbit-db-keystore')
+const type = 'orbitdb'
+let supportedTypes = {
+  'orbitdb': OrbitDBIdentityProvider,
+}
+
+const getHandlerFor = (type) => {
+  if (!IdentityProviders.isSupported(type)) {
+    throw new Error(`IdentityProvider type '${type}' is not supported`)
+  }
+  return supportedTypes[type]
+}
+
+class IdentityProviders {
+  constructor(options = {}) {
+    this._keystore = options.keystore || Keystore.create(options.keypath || './orbitdb/ipkeys')
+  }
+
+  async sign (identity, data) {
+    const signingKey = await this._keystore.getKey(identity.id)
+    if (!signingKey)
+      throw new Error(`Private signing key not found from Keystore`)
+
+    const signature = await this._keystore.sign(signingKey, data)
+    return signature
+  }
+
+  async verify (signature, publicKey, data) {
+    return this._keystore.verify(signature, publicKey, data)
+  }
+
+  async create(options = {}) {
+    options = Object.assign({}, { type }, options )
+    const IdentityProvider = getHandlerFor(options.type)
+    const identityProvider = new IdentityProvider(options)
+    const id = await identityProvider.createId(options)
+    const { publicKey, idSignature } = await this.signIdentity(id, options)
+    const pubKeyIdSignature = await identityProvider.signIdentity(publicKey + idSignature, options)
+    return new Identity(id, publicKey, idSignature, pubKeyIdSignature, IdentityProvider.type, this)
+  }
+
+  async signIdentity(id, options = {}) {
+    const keystore = options.keystore || this._keystore
+    const key = await keystore.getKey(id) || await keystore.createKey(id)
+    const publicKey = await key.getPublic('hex')
+    const idSignature = await keystore.sign(key, id)
+    return { publicKey, idSignature }
+  }
+
+  async verifyIdentity (identity, options = {}) {
+    const verified = await this._keystore.verify(
+      identity.signatures.id,
+      identity.publicKey,
+      identity.id
+    )
+    options = Object.assign({}, options, { provider: this })
+    return verified && await IdentityProviders.verifyIdentity(identity, options)
+  }
+
+  static async verifyIdentity(identity, options = {}) {
+    const IdentityProvider = getHandlerFor(identity.type)
+    return await IdentityProvider.verifyIdentity(identity, options)
+  }
+
+  static async createIdentity (options = {}) {
+    const identityResolver = new IdentityProviders(options)
+    return await identityResolver.create(options)
```

Comment:
The naming `identityResolver` here is a bit confusing as "resolver" would imply "go a find data for this" whereas we're actually creating the Identity here. So I think the variable here should be named `identityProvider` and the `create` function should be `createIdentity` (we're creating an Identity, whereas "IdentityProviders.create()" implies creating an IdentityProvider).